### PR TITLE
Make no-user entry with avatar when there is no others with access

### LINKED ISSUE #18875: Emptycontent view for "Others with access" section
If there are no others with access, on opening the "Others with access" menu, show an entry like a person entry saying:

> **No others with access**
> People with access to parent folders will show up here

(As discussed in https://github.com/nextcloud/server/pull/18701 )

### PR DESCRIPTION
**Pull request by JeaNugroho**
* Resolves: #18875

## Summary
Make an empty entry with predefined info for when there's no others with access. This can be seen in the share tab inside the files view by clicking on any file/folder share button.
### Before
[Link to conversation](https://github.com/nextcloud/server/pull/18701#issue-545934317)
### After (with avatar - first commit)
After-ToggleOthersWithAccess
### After (without avatar - second commit)
After-ToggleOthersWithAccess-woAvatar

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [x] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

**Review by jancborchardt — COMMENTED** [review]
I'd say without avatar looks better, but the indent is good. (So the emptycontent text is indented as if there is an avatar, do you know what I mean? :)

File: apps/files_sharing/l10n/en_GB.js
Comment (artonge) [inline apps/files_sharing/l10n/en_GB.js:1]: No need to touch l10n file, can you revert this change?

File: apps/files_sharing/src/components/SharingEntryInherited.vue [hunk 1]
```diff
@@ -21,28 +21,39 @@
...
 <template>
-	<SharingEntrySimple :key="share.id"
-		class="sharing-entry__inherited"
-		:title="share.shareWithDisplayName">
-		<template #avatar>
-			<NcAvatar :user="share.shareWith"
-				:display-name="share.shareWithDisplayName"
-				class="sharing-entry__avatar" />
+	<span>
+		<template v-if="!shares">
```
Comment (artonge) [inline apps/files_sharing/src/components/SharingEntryInherited.vue:25]: Where does `shares` come from? I do not think it exists.

File: apps/files_sharing/src/views/SharingInherited.vue [hunk 2]
```diff
@@ -132,10 +126,16 @@ export default {
...
-				this.shares = shares.data.ocs.data
-					.map(share => new Share(share))
-					.sort((a, b) => b.createdTime - a.createdTime)
+
+				if (shares.length > 0) {
+					this.shares = shares.data.ocs.data
+						.map(share => new Share(share))
+						.sort((a, b) => b.createdTime - a.createdTime)
+				} else {
+					this.shares.push(null)
```
Comment (artonge) [inline apps/files_sharing/src/views/SharingInherited.vue:135]: I would rather have a condition in the template around line 39 to render the text if `shares.length === 0`. I think this would also prevent the need of touching `SharingEntryInherited.vue`.

**Review by artonge — COMMENTED** [review]
Nice idea, few comments though :)

File: apps/files_sharing/src/components/SharingEntryInherited.vue (hunk 1 above)
Comment (JeaNugroho) [reply apps/files_sharing/src/components/SharingEntryInherited.vue:25]: @artonge Good eye! I meant `share` instead. Upcoming changes (commits) will have this update

File: apps/files_sharing/l10n/en_GB.js
Comment (JeaNugroho) [reply apps/files_sharing/l10n/en_GB.js:1]: @artonge Definitely! Do we just put string literals in the template itself, or do we not have a file of string constants somewhere to define this?

And to ask in the right time, is I10n file only for translation purposes? And is this string I'm working with not going to be translated?

**Comment by JeaNugroho** [comment]
@jancborchardt 


> I'd say without avatar looks better, but the indent is good. (So the emptycontent text is indented as if there is an avatar, do you know what I mean? :)

This is what it looks like since the 3rd commit:
Screen Shot 2023-06-27 at 7 50 42 PM
Screen Shot 2023-06-27 at 7 50 10 PM

I hope this is what we wanted to achieve 😁

File: apps/files_sharing/src/views/SharingInherited.vue [hunk 3]
```diff
@@ -25,7 +25,6 @@
 		<!-- Main collapsible entry -->
 		<SharingEntrySimple class="sharing-entry__inherited"
 			:title="mainTitle"
-			:subtitle="subTitle"
```
Comment (artonge) [inline apps/files_sharing/src/views/SharingInherited.vue:28]: Why is this removed?

File: apps/files_sharing/src/views/SharingInherited.vue [hunk 4]
```diff
@@ -92,11 +99,6 @@ export default {
 		mainTitle() {
 			return t('files_sharing', 'Others with access')
 		},
-		subTitle() {
-			return (this.showInheritedShares && this.shares.length === 0)
-				? t('files_sharing', 'No other users with access found')
-				: ''
-		},
```
Comment (artonge) [inline apps/files_sharing/src/views/SharingInherited.vue:99]: Why is this removed?

File: apps/files_sharing/src/views/SharingInherited.vue [hunk 5]
```diff
@@ -37,11 +36,19 @@
...
-			:key="share.id"
+		<div v-if="loaded && shares.length === 0" class="sharing-entry__desc">
+			<span class="sharing-entry__title">{{ 'No others with access' }}</span>
+			<p>
+				{{ 'People with access to parent folders will show up here' }}
+			</p>
+		</div>
+		<template v-else>
+			<SharingEntryInherited v-for="share in shares"
+			:key="share ? share.id : 0"
```
Comment (artonge) [inline apps/files_sharing/src/views/SharingInherited.vue:47]: What's the reason for that? `share` should be always defined with the new approach

File: apps/files_sharing/src/views/SharingInherited.vue [hunk 6]
```diff
@@ -37,11 +36,19 @@
 		</SharingEntrySimple>
 
 		<!-- Inherited shares list -->
-		<SharingEntryInherited v-for="share in shares"
-			:key="share.id"
+		<div v-if="loaded && shares.length === 0" class="sharing-entry__desc">
+			<span class="sharing-entry__title">{{ 'No others with access' }}</span>
+			<p>
+				{{ 'People with access to parent folders will show up here' }}
```
Comment (artonge) [inline apps/files_sharing/src/views/SharingInherited.vue:42]:
```suggestion
				{{ t('files_sharing', 'People with access to parent folders will show up here') }}
```

**Review by artonge — COMMENTED** [review]
Nice, feels much cleaner :)

File: apps/files_sharing/l10n/en_GB.js
Comment (artonge) [reply apps/files_sharing/l10n/en_GB.js:1]: See other comment

**Review by nimishavijay — APPROVED** [review]
Looks great with the fix! 🚀

File: apps/files_sharing/src/views/SharingInherited.vue (hunk 3 above)
Comment (JeaNugroho) [reply apps/files_sharing/src/views/SharingInherited.vue:28]: @artonge It was previously applied to make a quick patch as a feedback for 'no others with access'. We can see that discussion [here](https://github.com/nextcloud/server/pull/18701#issue-545934317).

File: apps/files_sharing/src/views/SharingInherited.vue (hunk 4 above)
Comment (JeaNugroho) [reply apps/files_sharing/src/views/SharingInherited.vue:99]: @artonge It was previously applied to make a quick patch as a feedback for 'no others with access'. We can see that discussion [here](https://github.com/nextcloud/server/pull/18701#issue-545934317).

File: apps/files_sharing/src/views/SharingInherited.vue (hunk 5 above)
Comment (JeaNugroho) [reply apps/files_sharing/src/views/SharingInherited.vue]: @artonge You're right. Have just committed a fifth commit to clean this up and just make it `:key="share.id` since `share` will most likely not be null 👌

File: apps/files_sharing/src/views/SharingInherited.vue (hunk 6 above)
Comment (JeaNugroho) [reply apps/files_sharing/src/views/SharingInherited.vue]: @artonge Newer changes kicked in and it should be good now 👍

File: apps/files_sharing/src/views/SharingInherited.vue [hunk 7]
```diff
@@ -37,11 +36,19 @@
 		</SharingEntrySimple>
 
 		<!-- Inherited shares list -->
-		<SharingEntryInherited v-for="share in shares"
+		<div v-if="loaded && shares.length === 0" class="sharing-entry__desc">
+			<span class="sharing-entry__title">{{ 'No others with access' }}</span>
```
Comment (artonge) [inline apps/files_sharing/src/views/SharingInherited.vue:40]: Sorry, missed this label. This need the same pattern as the other one:

```suggestion
			<span class="sharing-entry__title">{{ t('files_sharing', 'No others with access') }}</span>
```

Comment (JeaNugroho) [reply apps/files_sharing/src/views/SharingInherited.vue]: @artonge  Should be good now!

**Comment by JeaNugroho** [comment]
> Looks great with the fix! 🚀

@nimishavijay 

Thanks, appreciate it! Do you mind reviewing another PR that I made? It's solving a simpler issue, and you can find it [here](https://github.com/nextcloud/server/pull/38952).

On another note, I'm looking to work on some backend issues, and I would appreciate it if I can reach out to someone who I can ask about backend or any other questions to. And I might as well ask where would be a good place to chat with the person (if it's through help.cloud.com, I'd like to know which channel. Thanks in advance!

**Review by artonge — APPROVED** [review]

**Comment by JeaNugroho** [comment]
How can I merge these changes? @artonge @nimishavijay @jancborchardt 
It says "Required statuses must pass before merging", but I'm not exactly sure on how to fix them or if there's any directions to do it

**Comment by nimishavijay** [comment]
@JeaNugroho I am not too sure about the failing checks, maybe @szaimen or @artonge can help out here?
help.nextcloud.com is a great place to start asking about development questions! We also have a [community Talk chat](https://cloud.nextcloud.com/call/xs25tz5y) where there are a lot of active contributors who will also be able to help you out :)

**Comment by susnux** [comment]
> It says "Required statuses must pass before merging", but I'm not exactly sure on how to fix them or if there's any directions to do it.

You need to fix the issues detected by the CI:
1. The Node CI fails because you need to build the JS assets and commit them (`npm run build`)
2. The Lint CI fails because you have some code style issues (they are also reported here on the changed files section), most of the time you can solve it by run `npm run lint:fix`.

**Review by susnux — APPROVED** [review]
LGTM, just see the comment about resolving the CI issues :)

**Comment by artonge** [comment]
Hey @JeaNugroho can you rebase and commit the compiled files again?
Any conflict under /dist can be ignored.

**Comment by JeaNugroho** [comment]
> Hey @JeaNugroho can you rebase and commit the compiled files again?
> Any conflict under /dist can be ignored.

@artonge Working on it! Is that why Node CI failed even after I tried to fix it? 😅

**Comment by JeaNugroho** [comment]
Have just rebased! I assume I'll have to wait for all checks to complete and continue fixing more issues in case some required ones fail. I'll need more guidance there as well

**Comment by artonge** [comment]
Can you run `npm run lint:fix`, and then rebase, compile, commit again?

Something like:

```shell
npm run lint:fix
git fetch
git rebase origin/master
npm run build
git commit -am '...' --signoff
git push
```

**Comment by JeaNugroho** [comment]
@artonge @susnux 
I'm getting this after `npm run lint:fix`

Screen Shot 2023-07-04 at 8 17 06 AM

How can I fix this issue?

**Comment by artonge** [comment]
Ah, maybe a dependency issue. Can you run `npm ci` and try again

```shell
git fetch
git rebase origin/master
# Complete the rebase

npm ci
npm run lint:fix
npm run build
git commit -am '...' --signoff
git push
```

**Comment by JeaNugroho** [comment]
> Ah, maybe a dependency issue. Can you run `npm ci` and try again:
> 
> ```shell
> git fetch
> git rebase origin/master
> # Complete the rebase
> 
> npm ci
> npm run lint:fix
> npm run build
> git commit -am '...' --signoff
> git push
> ```

@artonge I already ran `nom install --save-dev @nextcloud/eslint-config`. Is that okay or is it going to cause trouble? I'll try to run that `npm ci`. I've tried running the sequence of commands up until `rpm run build, but am going to restart from `npm ci`

**Comment by JeaNugroho** [comment]
@artonge The commit command you wrote has a `-am` in it. Is that "a" supposed to be there?

**Comment by JeaNugroho** [comment]
@artonge I'm getting this. I'm not sure how to fix this git push issue. Fyi, my origin is the forked version of the repo, while my upstream is Nextcloud's actual server repo

Screen Shot 2023-07-04 at 8 59 16 AM

**Comment by artonge** [comment]
Seems like your local state is behind the remote one. You might have deleted a commit locally.
If you are sure about you, you can `git push --force-with-lease`

**Comment by JeaNugroho** [comment]
> Seems like your local state is behind the remote one. You might have deleted a commit locally.
> If you are sure about you, you can `git push --force-with-lease`

@artonge Done! We'll see if this works out, hopefully it does =)

**Comment by artonge** [comment]
Please revert any change under `/3rdparty`

`git submodule update`

**Comment by JeaNugroho** [comment]
@artonge So based on your last two comments, what should I do and what commands should I run? I only ran the commands discussed in this PR so far

**Comment by artonge** [comment]
This page https://github.com/nextcloud/server/pull/39022/files need to display your changes only and the result of `npm run build`.

The rebase might have gone wrong.
If creating another PR is easier for you, then do that. But it should be possible to revert the state of this PR to contain your change only.

**Comment by skjnldsv** [comment]
Lots of conflicts and unrelated changes
